src/device_handle.rs

```diff
@@ -540,22 +540,19 @@ impl<T: UsbContext> DeviceHandle<T> {
     pub fn read_string_descriptor_ascii(&self, index: u8) -> crate::Result<String> {
         let mut buf = Vec::<u8>::with_capacity(255);
 
-        let buf_slice = unsafe { slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.capacity()) };
-
-        let ptr = buf_slice.as_mut_ptr() as *mut c_uchar;
-        let len = buf_slice.len() as i32;
+        let ptr = buf.as_mut_ptr() as *mut c_uchar;
+        let len = buf.len() as i32;
 
         let res =
             unsafe { libusb_get_string_descriptor_ascii(self.handle.as_ptr(), index, ptr, len) };
 
         if res < 0 {
             return Err(error::from_libusb(res));
+        } else if res > len {
+            return Err(Error::Other);
         }
 
-        unsafe {
-            buf.set_len(res as usize);
-        }
-
+        buf.resize(res as usize, 0u8);
```

Comment:
It's wrong. Resize fill all vec with zero.
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize

Reply:
No, I don't believe so. The value 0u8 is used to fill the new space if the vector grows. We already checked in the line above that the vector can't grow, so it's meaningless in this particular case.

Check the docs:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize

> If new_len is greater than len, the Vec is extended by the difference, with each additional slot filled with value. If new_len is less than len, the Vec is simply truncated.

But, I guess, really I should test it. Give me a minute...

Reply:
Yeah. it works...
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0b983877d5717adf0fb3e40e7f52a558

```
fn main() {
    let mut v = vec![ 5u8; 4 ];
    println!("{:?}", v);
    
    // Grow larger
    v.resize(6, 1u8);
    println!("{:?}", v);

    // Shrink smaller
    v.resize(3, 0u8);
    println!("{:?}", v);
}
```
Prints:
```
[5, 5, 5, 5]
[5, 5, 5, 5, 1, 1]
[5, 5, 5]
```

Reply:
Please write correct test.
We alloc with Vec::<u8>::with_capacity(255)

Reply:
Sure. 

But I just looked at the source code for Vec. If new_len < len, then it just calls the _truncate()_ function, which is a public function. So we could just call that one directly. I'll update the PR.

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate

Here's the source. Clearly it is not changing the remaining contents of the Vec:
```
    pub fn truncate(&mut self, len: usize) {
        unsafe {
            if len > self.len { return; }
            let s = self.get_unchecked_mut(len..) as *mut _;
            self.len = len;
            ptr::drop_in_place(s);
        }
    }
```

Reply:
Oh... wait. Now I see what you're saying....
Ha. Yes. OK! I'm wrong.